src/kernel/mempool_removal_reason.h

```diff
@@ -5,20 +5,77 @@
 #ifndef BITCOIN_KERNEL_MEMPOOL_REMOVAL_REASON_H
 #define BITCOIN_KERNEL_MEMPOOL_REMOVAL_REASON_H
 
+#include <primitives/transaction.h>
+#include <uint256.h>
+
 #include <string>
+#include <variant>
 
 /** Reason why a transaction was removed from the mempool,
  * this is passed to the notification signal.
  */
-enum class MemPoolRemovalReason {
-    EXPIRY,      //!< Expired from mempool
-    SIZELIMIT,   //!< Removed in size limiting
-    REORG,       //!< Removed for reorganization
-    BLOCK,       //!< Removed for block
-    CONFLICT,    //!< Removed for conflict with in-block transaction
-    REPLACED,    //!< Removed for replacement
+
+struct ExpiryReason {
+    std::string toString() const noexcept
+    {
+        return "expiry";
+    }
+};
+
+struct SizeLimitReason {
+    std::string toString() const noexcept
+    {
+        return "sizelimit";
+    }
+};
+
+struct ReorgReason {
+    std::string toString() const noexcept
+    {
+        return "reorg";
+    }
+};
+
+struct BlockReason {
+    std::string toString() const noexcept
+    {
+        return "block";
+    }
+};
+
+struct ConflictReason {
+    uint256 conflicting_block_hash;
+    unsigned int conflicting_block_height;
+
+    explicit ConflictReason(const uint256& conflicting_block_hash, int conflicting_block_height) : conflicting_block_hash(conflicting_block_hash), conflicting_block_height(conflicting_block_height) {}
```

Comment:
In 4c64dd5a3cc4a48e5470fff26c99dde20f81c7e0   "Change MemPoolRemovalReason to variant type"
nit:  this line is too long, please break.